### ISSUE TITLE
No ability to check if `ng-content` was provided to the component

### ISSUE THREAD

**Issue by sharikovvladislav** [angular/angular#26083]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior

We can use `ng-content` in our components. But we don't have any options out of the box to check if anything is passed as `ng-content` or not. Actually, was anything passed between `>` and `</` of our component usage in HTML.

We can do this by adding some wrap component (and if I understood correctly you can not use stuff like `ng-container` there). Like div. And you put ng-content to it. You can get ref for this div and check if this div has children. If yes - ng-content was provided, otherwise - no ng-content. Huge minus of this approach is you have to create some additional entity to check it.

## Expected behavior

We can create some util function for that. You will have ability to just call that util and detect if `ng-content` was passed.

## Minimal reproduction of the problem with instructions

This is not a bug. This example shows us how we can do this at the moment. https://stackblitz.com/edit/angular-xdirt8

## What is the motivation / use case for changing the behavior?

We will not be forced to add some wrappers and it will be easier to check this stuff.

## Environment

<pre><code>
Angular version: 6.0.0

Others:

</code></pre>

I tried to ask it on stackoverflow and I am not only one person who is wondered how to do this.

https://stackoverflow.com/questions/52479608/how-to-check-if-ng-content-was-passed-into-the-component
https://stackoverflow.com/questions/35107211/in-angular-2-how-to-check-whether-ng-content-is-empty
https://stackoverflow.com/questions/38692881/how-to-check-whether-ng-content-exists

**Comment by sharikovvladislav** [comment]
Any ideas on this? Maybe I am missing something and there is something what I can use?

**Comment by mlc-mlapis** [comment]
@sharikovvladislav ... you can always mark the transcluded content and later use `@ContentChild('mark')` as ...
```
<my-parent>
   <my-child #mark>
      ...
   <my-child>
</my-component>
```

**Comment by sharikovvladislav** [comment]
I know this. This is not what I ask.

I have a bit different situation. I don't know what will be projected to my component. And I don't want everybody to pass something to my component with `mark` ref. It requires some additional arrangements in team. Not sure this is true way. But yes, this will work.

**Comment by sharikovvladislav** [comment]
@mlc-mlapis It is not explicit API at all.

> If somebody has access to the source code, anything can be passed.

It is not good abstraction if you need to read source codes of some stuff to use it. 

My problem can be solved by your way or by my suggestion (well, not my, I think I found this somewhere about half a year ago), but it is not as explicit and as easy as using simple util function (or another simple API).

**Comment by kemsky** [comment]
Similar solution to yours:
```
<ng-container><span #ref><ng-content></ng-content></span></ng-container>
<ng-container *ngIf="ref.childNodes.length == 0">{{name || title}}</ng-container>
```

For now the best way is probably to use templates instead of projections (not perfect though).

**Comment by leomendoza123** [comment]
The problem is that if there is an `*ngIf` evaluated to false inside the `ng-container`, Angular is going to add a comment (at least on development mode) like 
```
bindings={
  "ng-reflect-ng-if": null
}-->
```

And that's is going to be counted as  a node, so ` *ngIf="ref.childNodes.length == 0"` is not going to work

**Comment by leomendoza123** [comment]
So if you are projecting any *ngIf inside a `ng-content` and you want to check if the projection is empty.

You can do this

```
<div #contentProjection>
   <ng-content></ng-content>
</div>`
<div *ngIf="hasContentProjection()"> NO CONTENT PROJECTION</div>
```

```
@ViewChild('contentProjection') contentProjection: ElementRef

hasContentProjection() {
    const detailContentProjection: NodeList = (this.contentProjection
      .nativeElement as HTMLElement).childNodes
    for (let x = 0; x < detailContentProjection.length; x++) {
      console.log(detailContentProjection.item(x).nodeType)
      if (detailContentProjection.item(x).nodeType !== 8) {
        return true // nodeType == 8 means is a comment
      }
    }
 }
```

**Comment by mlc-mlapis** [comment]
@leomendoza123 ... sure, it would work, but this is much more simpler way: https://github.com/angular/angular/issues/26083#issuecomment-424006120

**Comment by trotyl** [comment]
Also, there's a `cdkObserveContent` directive in the CDK should be able to help with this situation: https://material.angular.io/cdk/observers/overview

**Comment by imike57** [comment]
Still not a true solution for this ? Seem to be a basic usage... absolutely required when building a library !

**Comment by ScottMGerstl** [comment]
Consider looking into the `:empty` CSS Selector

**Comment by alberto-chiesa** [comment]
> 
> 
> @leomendoza123 ... sure, it would work, but this is much more simpler way: [#26083 (comment)](https://github.com/angular/angular/issues/26083#issuecomment-424006120)

No, it's not simpler: in your solution, you require any user of the component to add a div with a magic id. In the solution of @leomendoza123  you put the complexity in the reusable component.

Take this with a grain of salt, but when designing a library, it doesn't really matter how much complexity you have in the implementation. The only thing that matters is ease of use (given the same feature set).

**Comment by joshcomley** [comment]
This is a basic feature that should exist. Current solutions of counting child nodes are explicitly web-based and do not reflect the fact Angular can be used in many other scenarios, and has been built from the ground up to be able to do so. I am using NativeScript Angular, for example, and css empty selector or counting child nodes won't necessarily suffice - this should be included as standard.

**Comment by stepancar** [comment]
@sharikovvladislav 
https://medium.com/@istepancar/how-to-avoid-render-wrapper-for-ng-content-when-it-not-passed-in-dartangular-d89cb3b026a

**Comment by muuvmuuv** [comment]
Sad thing that the CSS trick with `:empty` does not work any more...

```html
```
![Bildschirmfoto 2020-05-11 um 12 25 57](https://user-images.githubusercontent.com/11534760/81551697-926efc00-9382-11ea-97d1-784e7c8b1535.png)

**Comment by Adit-COCO-Garg** [comment]
I agree with @joshcomley. One shouldn't have to write complex angular code or use some hacky stuff to figure out if any content is being passed in. It should be a simple boolean variable available to the template. React may be of some inspiration. 

Commenting to push this issue up and also sharing that this is a much-needed feature.

**Comment by klemenoslaj** [comment]
This can be particularly painful in case you would like to keep the html semantics and styling internal to the component but only accept text and/or children.

Like:

```html
<h3 class="headline"><ng-content select="headline"></ng-content></h3>
```

I you'd want to keep `h3` and class `headline` internal for the component, you can end up with empty `h3` element which can cause accessibility issues.

**Comment by niki-lijnen** [comment]
From What I understand, the request here is to have something similar as what was done in AngularJS where you could query the $transclude service with 

`$transclude.isSlotFilled('mySlotName')`

I have used this a number of times in AngularJS to add an expand button to a panel. I definitely miss this in Angular.

**Comment by zolakt** [comment]
Content projection in Angular is incredibly lacking behind every other major framework. Sadly even compared to AngularJS.

The use case shown here is a perfect example 
https://github.com/angular/angular/issues/26083#issuecomment-964115189

Is there even any kind of workaround for this, as ugly as it may be?
Having empty elements hanging around is just not acceptable, for a wide variety of reasons.

**Comment by mohaiminalaoun** [comment]
Not sure if this is being worked on, but it definitely should not require developers to write hacky code to get this to work. It seems like a very simple case that would be covered by the angular framework.

**Comment by gtteamamxx** [comment]
That should be angular 16 feature :)

**Comment by spicy-tomato** [comment]
> That should be angular 16 feature :) @up

Should be 16.1 feature, hope so...

**Comment by RazeiXello** [comment]
Thanks for your contributions everyone. The solution I went with was something like this:

```html
<ng-container><span #ref><ng-content></ng-content></span></ng-container>
<ng-container *ngIf="ref.children.length == 0">{{name || title}}</ng-container>
```

Which also ignores this comment:

```
bindings={
  "ng-reflect-ng-if": null
}-->
```
Quote from [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Node/childNodes)
> It is important to keep in mind that childNodes includes all child nodes, including non-element nodes like text and comment. To get a collection containing only elements, use [Element.children](https://developer.mozilla.org/en-US/docs/Web/API/Element/children) instead.

Hoping we have a way of doing this and [default content](https://github.com/angular/angular/issues/12530) in a simpler way in the future.

**Comment by haskelcurry** [comment]
Any updates on this?

**Comment by Erbenos** [comment]
Curious after reading @joshcomley's comment, I was doing some research into a renderer agnostic way emptyness check, hoping that reading some higher level generated information would be sufficient.

First, I came up with this:

https://stackblitz.com/edit/stackblitz-starters-ajeiyh

Which is basically is basically creating a view out of templateRef that wraps ng-content, and then reading the nodes produced by the renderer.

```
<ng-template #contentTpl><ng-content/></ng-template>
  <ng-container *ngIf="isFilled; else defaultTpl">
    <ng-container *ngTemplateOutlet="contentTpl" />
  </ng-container>
  <ng-template #defaultTpl>
    Default Tpl
  </ng-template>
```

```
  public ngOnInit(): void {
    this.view = this.vcr.createEmbeddedView(this.contentTpl);
  }

  public ngAfterContentChecked(): void {
    this.isFilled = Boolean(
      // Comment from '@angular/compiler'
      this.view?.rootNodes?.filter((node) => !(node instanceof Comment)).length
    );
    this.cdr.markForCheck();
  }
```

Trying to do the same within NativeScript application however was not fruitful, cause even the rootNodes item types are completely different, though not surprisingly so as NativeScript doesn't render into DOM. Such solution could be tweaked however:

https://stackblitz.com/edit/nativescript-stackblitz-templates-ambzcc

```
  public ngAfterContentChecked(): void {
    this.isFilled = Boolean(
      // InvisibleNode from '@nativescript/angular'
      this.view?.rootNodes?.filter((node) => !(node instanceof InvisibleNode))
        .length
    );
    this.cdr.markForCheck();
  }
```

Do note however that what is "empty" is debatable. Eg.: the DOM version does not classify the textless text nodes as empty, while the Nativescript's one does.

Really wish this was easier and unified somewhat.

**Comment by maxfriedmann** [comment]
Here's a signal based version of @Erbenos one:

```html
<ng-template #contentTemplate><ng-content></ng-content></ng-template>
@if (!hasNgContent()) {
  <span>doSomething</span>
}
```

```typescript
  private vcr = inject(ViewContainerRef);
  private contentTemplate = viewChild<TemplateRef<unknown>>("contentTemplate");
  protected ngContentView = computed(() => this.vcr.createEmbeddedView(this.contentTemplate()));
  protected hasNgContent = computed(() => {
    const view = this.ngContentView();
    return view?.rootNodes?.filter((node) => !(node instanceof Comment)).length;
  });
```

Still creates an embedded view of `ng-content`, no matter if you call `hasNgContent()` or not.

**Comment by destus90** [comment]
Seems like with this merged PR https://github.com/angular/angular/pull/54854 in v18 we will be able to do it easier 🎉

**Comment by muuvmuuv** [comment]
This fixes one case, what about programmatic access?

I really wished for something like (control-flow):

```html
<!-- only display projection if passed -->
@content (".selector")

@content (".selector") {
	<!-- display while no content is injected -->
}
```

```ts
class ViewComponent {
	// Programmatically access content view template
	readonly content = viewContent(".selector")
	// Mark required so a child must be passed
	readonly content = viewContent.required(".selector")
	// Access multiple content projections
	readonly content = viewContents(SomeOtherComponent)

	// Do some Signal magic afterwards
	readonly hostClassNoChildren = hostBinding("class.no-children", computed(() => {
		return this.content.length === 0
	})
}
```

**Comment by GertrudMathilde** [comment]
I'd love to use a feature like this to change CSS classes like [class.whatever]="ngcontent.isEmpty". I'm using childNodes.length for now. children.length doesn't work if there's only text with no wrapping Element.

**Comment by netoctone** [comment]
Thanks @destus90, indeed since angular 18 it's possible to specify fallback content for `ng-content` https://medium.com/@davidepassafaro/angular-fallback-content-in-ng-content-5aceb31acc0a#c6d6
which partially covers this issue.
